### PR TITLE
Neural Net

### PR DESCRIPTION
**Pull request by USSiamaboat**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 env
 __pycache__
 *.pt
-*.pkl
\ No newline at end of file
+*.pkl
+.idea/
```
Comment (USSiamaboat) [inline .gitignore]: Do not add IDE specific things to master, consider adding to .gitignore

File: Dataset.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import imported_game
+import numpy as np
+import os
+from data_preprocess import node_to_tensor
+
+
+def human_target_policy(gameNode):
+    """ return a board array with a 1 where the human player moved"""
```
Comment (USSiamaboat) [inline Dataset.py]: Why is there a space after the first `"""` but not before the second `"""`

File: Dataset.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import imported_game
+import numpy as np
+import os
+from data_preprocess import node_to_tensor
+
+
+def human_target_policy(gameNode):
```
Comment (USSiamaboat) [inline Dataset.py]: Consider using PyTorch built-in one-hot encoding functions instead of manually implementing

File: dataset.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import imported_game
+import numpy as np
+import os
+from data_preprocess import node_to_tensor
+
+
+def human_target_policy(gameNode):
```
Comment (USSiamaboat) [inline dataset.py:7]: Consider using PyTorch built-in one-hot encoding functions

File: dataset.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    """ return a board array with a 1 where the human player moved"""
```
Comment (USSiamaboat) [inline dataset.py]: Spacing is inconsistent, capitalization?

File: dataset.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    # handle move
+    else:
+        row = gameNode.prev_move[0]
+        col = gameNode.prev_move[1]
+        target[gameNode.size * row + col] = 1
+
+    return target
+
+
+class Dataset:
```
Comment (USSiamaboat) [inline dataset.py:24]: Consider adding a docstring to this class

File: network.py [hunk 7]
```diff
@@ -60,47 +63,129 @@ def forward(self, x):
...
+        # 2. Batch Normalisation on the single channel output
+        self.bn = nn.BatchNorm2d(1)
+
+        # 3. A rectifier non-linearity
+        self.relu = nn.ReLU()
+
+        # 4. A fully connected linear layer to a hidden layer of size 256
+        self.fc = nn.Linear(board_size * board_size, 256)
+
+        # 5. ReLu again
```
Comment (onatozer) [inline network.py:112]: There's a comment saying to relu but no code that does it

**Review by onatozer — CHANGES_REQUESTED** [review]

File: dataset.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import imported_game
```
Comment (USSiamaboat) [inline dataset.py]: Consider using `from imported_game import ...`

File: dataset.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    def load_games(self, game_directory):
+        for i, game_file in enumerate(dirs := os.listdir(game_directory)):
+            print(f"[{i}/{len(dirs)}] Loading game {game_file}")
+            filepath = os.path.join(game_directory, game_file)
+            this_game = imported_game.ImportedGame(filepath)
+
+            # parse through the game
+            node = this_game.linked_list()
+            last_human_policy = human_target_policy(node)
+            while (node := node.prev) is not None:
```
Comment (USSiamaboat) [inline dataset.py]: Walrus operator is not good practice

File: dataset.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+
+
+
+if __name__ == '__main__':
+    dataset = Dataset("games")
+
+    print(len(dataset.positional_data))
+
+    # Do not print this, way too long
```
Comment (USSiamaboat) [inline dataset.py]: Remove before merge with master

File: game_node.py [hunk 11]
```diff
@@ -103,11 +106,14 @@ def create_child(self, loc: tuple[int, int]) -> Self:
...
             board = board.create_child((row, col))
+            print(board.get_game_data())
+            print(board.get_game_data().shape)
+
         except KeyboardInterrupt:
             print("\nKeyboard Interrupt. Game Ended")
             break
-        except:
-            print("Error while processing move. Try again.")
+        # except:
```
Comment (USSiamaboat) [inline game_node.py:115]: Remove before merge with master

File: network.py (hunk 7 above)
Comment (USSiamaboat) [reply network.py:112]: We already have ReLU in 3, should be fixed anyways because comment is not clear about that

File: supervised_learning.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import torch
+from Dataset import Dataset
```
Comment (USSiamaboat) [inline supervised_learning.py]: Update for `Dataset.py` name change to `dataset.py`

File: supervised_learning.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+        return value_loss + policy_loss
+
+
+def train(model, epochs=5, batch_size=32):
+    dataset = Dataset("games")
+
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, shuffle=True)
+
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
```
Comment (USSiamaboat) [inline supervised_learning.py]: Make the learning rate adjustable

File: supervised_learning.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+            pi_hat, z_hat = model(s)
+
+            loss = criterion(z.to(torch.float32), z_hat, pi, pi_hat)
+            loss.backward()
+
+            optimizer.step()
+
+            print(f"Epoch {epoch}, Batch {i}: Loss: {loss.item()}")
+    return
```
Comment (USSiamaboat) [inline supervised_learning.py]: Why

**Review by USSiamaboat — COMMENTED** [review]
Add config also

**Review by USSiamaboat — COMMENTED** [review]
Looks good
